### PR TITLE
Compute `max_fee_per_gas` Per Driver Run

### LINKED ISSUE #1360: Compute `max_fee_per_gas` Upfront
We should compute the `max_fee_per_gas` upfront instead of when bumping gas:

* This sets an upper bound on total gas spend (that we can check at simulation time)
* We don't need the `max_fee_per_gas` to be so much higher than the `effective_gas_price` as we near the end of the deadline (since the actual `base_gas` can't increase so much near the end).
* We can submit transactions with increasing gas prices by only bumping the max priority fee
* No need to re-estimate gas for subsequent transactions during a settlement transaction execution.

The main advantage of this approach is that we don't need to have such a high upfront balance in the solver accounts to submit the transaction, while still allowing for a maximally large gas spike.

### PR DESCRIPTION
**Pull request by nlordell**
Fixed #1360 

This PR changes the driver run loop to compute a `max_fee_per_gas` value upfront. This has a few benefits, namely:
- We don't over-estimate a `max_fee_per_gas` value at solver competition time. Specifically, if we take a gas price estimate of `max_fee_per_gas: 100, base_fee_per_gas: 50` and scale it by a maximum factor (lets say 5 for argument's sake), we would end up with an EIP-1559 estimate of `max_fee_per_gas: 500, base_fee_per_gas: 250`. This would require a solver to have `500 * gas_limit` balance in order for their transaction to be valid. However, if we assume that the `base_fee_per_gas` would not increase by more than a factor of `5`, then we actually only need `max_fee_per_gas: 250, base_fee_per_gas: 250` for estimation - meaning that a solver only needs half the balance for their transaction to be valid, making estimation more efficient. This is needed because the submission loop does gas estimates internally, and there is a constant factor between `base_fee_per_gas` and `max_fee_per_gas`.
- We can verify a solver's balance at the start of a driver run-loop, and be guaranteed that it is sufficient for the entire submission period.

### Test Plan

Added some unit tests to verify the `max_fee_per_gas` logic.

### CODE REVIEW

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 1]
```diff
@@ -77,3 +80,51 @@ impl Verified {
...
+}
+
+impl Gas {
+    /// Computes settlement gas parameters given estimates for gas and gas
+    /// price.
+    pub fn new(estimate: eth::Gas, price: eth::GasPrice) -> Self {
+        // Compute an upper bound for `max_fee_per_gas` for the given
+        // settlement. We multiply a fixed factor of the current base fee per
+        // gas, which is chosen to be the maximum possible increase to the base
+        // fee per gas over 10 blocks.
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/settlement.rs:104]: I think it makes sense to name the `1.125` base fee scaling constant here and write out the formular how you arrived at the `MAX_FEE_FACTOR`. Maybe something like:
```rust
/// `MAX_FEE_FACTOR` = `MAX_GAS_INCREASE_PER_BLOCK` ^ `SUBMISSION_DEADLINE_IN_BLOCK`
///                  = 1.125 ^ 10
```
Also some link explaining why `1.125` makes sense would be nice.

**Review by MartinquaXD — APPROVED** [review]
Looks good overall. Mostly nits but I think the explanation why `3.25` makes sense needs more information.

File: crates/driver/src/domain/competition/solution/settlement.rs [hunk 2]
```diff
@@ -77,3 +80,51 @@ impl Verified {
...
+        const MAX_FEE_FACTOR: f64 = 3.25;
+        let price =
+            eth::U256::from_f64_lossy(eth::U256::to_f64_lossy(price.base.into()) * MAX_FEE_FACTOR)
+                .into();
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/competition/solution/settlement.rs:108]: This is for an upper-bound, I don't think loss-less behavior is important here (as we will pay the `base_fee_per_gas` regardless of this value).

File: crates/driver/src/domain/eth/gas.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
+use {
+    super::{Ether, U256},
+    std::ops,
+};
+
+/// Gas amount.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct Gas(pub U256);
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/eth/gas.rs]: No, these are gas units. I will improve the doc-comment.

File: crates/solver/src/driver/gas.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    /// `max_fee_per_gas` that is allowed for any given run. This allows
+    /// settlement simulation to know upfront the actual minimum balance that
+    /// would be required by a solver account (since, as per EIP-1559, an
+    /// account needs at least `max_fee_per_gas * gas_limit` for a transaction
+    /// to be valid, regardless of the `effective_gas_price` the transaction is
+    /// executed with). This way, we:
+    /// - Don't need to over-estimate a `max_fee_per_gas` value to account for
+    ///   gas spikes, meaning we won't disregard solvers with lower, but
+    ///   sufficient balances
+    /// - Will only ever chose a solver IIF it will have enough balance to
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/solver/src/driver/gas.rs:27]: 🤦

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 5]
```diff
@@ -166,8 +166,13 @@ impl Solution {
         let access_list = simulator.access_list(tx.clone()).await?;
         let tx = tx.set_access_list(access_list.clone());
 
-        // Finally, get the gas for the settlement using the full access list.
-        let gas = simulator.gas(tx).await?;
+        // Third, get the gas parameters for the settlement using the full
+        // access list.
+        let gas = {
+            let estimate = simulator.gas(tx).await?;
+            let price = eth.gas_price().await?;
```
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/competition/solution/mod.rs:173]: `try_join!` doesn't do automatic error conversions with `From` 😮.

[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/competition/solution/mod.rs:173]: I left this out for now for two reasons:
1. It made the code uglier (subjective)
2. It doesn't have any real effect because of how `NativeGasEstimator` works under the hood (it updates it internal values in a background task, so reading the estimate isn't really blocking).

[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/driver/src/domain/competition/solution/mod.rs:173]: > it updates it internal values in a background task, so reading the estimate isn't really async

Didn't know that. Then I agree that we should keep the nicer looking code.

File: crates/driver/src/domain/competition/solution/settlement.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nlordell) [reply crates/driver/src/domain/competition/solution/settlement.rs:116]: When we execute a settlement, we specify that the `max_fee_per_gas` cannot exceed `settlement.gas.price`:

https://github.com/cowprotocol/services/blob/ef9c8d278dd54f1f96d42b4c1b64634c67fd7b9e/crates/driver/src/boundary/mempool.rs#L119-L130

And in the submission logic:

https://github.com/cowprotocol/services/blob/ef9c8d278dd54f1f96d42b4c1b64634c67fd7b9e/crates/solver/src/settlement_submission/submitter.rs#L164-L171

This means that, even if we run into a situation where:
1. We get an initial gas price estimate with `base: 10, max: 20`
2. We compute the maximum we want to pay as `base * factor ~= 33`
3. Start submission loop capping gas price to `33`
    - If the estimator gives us `base: 20, max: 40`, we will use `max: 33` (and be guaranteed to have enough balance for it).
    - If the estimator gives us `base: 100, max: 200`, we will fail the submission as the base fee surpassed our cap (this can happen, for example, on Gnosis Chain where the block time is shorter, because of exponents).
    
Basically this change makes us no longer rely on the estimator's `max_fee_per_gas` but to compute our own based on the current base fee and how much we reasonably want to over-pay what it currently is.

[the remark this answers is not in the document]
Comment (fleupold) [reply crates/driver/src/domain/competition/solution/settlement.rs:116]: I see. Thanks for the explanation
